File: py3/141.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """

        if head is None:
            return False

        p_slow, p_fast = head, head.next

        while p_fast is not None and p_fast != p_slow:
            p_fast = p_fast.next
            if p_fast is not None:
                p_fast = p_fast.next
            else:
                break
            p_slow = p_slow.next

        if p_fast is None:
            return False
        else:
            return True

File: go/src/141.go
func hasCycle(head *ListNode) bool {
	if head == nil {
		return false
	}

	fast := head.Next
	slow := head

	for fast != nil && slow != nil {
		if fast != slow {
			if fast.Next != nil {
				fast = fast.Next.Next
			} else {
				return false
			}
			slow = slow.Next
		} else {
			return true
		}
	}

	return false
}